### PR TITLE
#872: add encoding member bits in handler field

### LINKED ISSUE #872: Encode member collection handler types in the bit pattern for the handler
**What Needs to be Done?**

Right now, in {send,broadcast}UntypedHandler we pass around a boolean `member` to distinguish the handler type (member vs. function) on the send and receive side after the handler is registered. That bit can easily fit in the 64-bits for `HandlerType`. Encode it, remove it from messages, and from the interface.

### PR DESCRIPTION
**Pull request by jstrzebonski**
* adds encoding member bit in handler field
* cleans up all the places that previously used `member` boolean
* fixes handlers' ID extraction
* standardizes making and getting handlers
* remove redundant const refs to HandlerType

Fixes #872

### CODE REVIEW

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/pipe/pipe_manager_tl.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6365828&bid=20658976)

**Comment by jstrzebonski** [comment]
OK, so I found a problem, and it's pretty dumb :- |
`makeHandler()` accepts several boolean, some of them with default values. Adding another one, without taking care of correctly pass it to function, screwed up the code. I'm fixing it right know and adding utest to check such scenario.

Also I was thinking, maybe it might make sense to replace that parameters with structs that have only one explicit ctor? That way it won't be possible anymore to pass arguments in wrong order. Something like this:

```
struct IsAuto {
  constexpr explicit IsAuto(bool const is_auto) : is_auto_{is_auto} {}
  operator bool() const { return is_auto_; }

private:
  bool is_auto_;
};
```

and refactor makeHandler to accept args of that special types, instead of regular bools.

`HandlerIdentifierType` and `HandlerControlType` could also be refactor that way..

File: src/vt/handler/handler.cc [hunk 1]
```diff
@@ -68,8 +69,9 @@ namespace vt {
   vt_debug_print(
     handler, node,
     "HandlerManager::makeHandler: is_functor={}, is_auto={}, is_objgroup={},"
-    " id={:x}, control={:x}, han={:x}, is_trace={}\n",
-    is_functor, is_auto, is_objgroup, id, control, new_han, is_trace
+    "is_memer={} id={:x}, control={:x}, han={:x}, is_trace={}\n",
```
Comment (JacobDomagala) [inline src/vt/handler/handler.cc:72]: is_member typo ;)

**Review by JacobDomagala — COMMENTED** [review]
Looks good overall. After the fixes to CI it should be good to be merged :)

File: src/vt/registry/auto/auto_registry.cc [hunk 2]
```diff
@@ -52,77 +52,67 @@
...
-    return getAutoRegistryGen<ContType>().at(han_id).theTraceID();
-    break;
+    return getTraceID<AutoActiveObjGroupContainerType>(handler);
   }
+
   case RegistryTypeEnum::RegVrtCollection: {
-    using ContType = AutoActiveCollectionContainerType;
-    return getAutoRegistryGen<ContType>().at(handler).theTraceID();
-    break;
+    return getTraceID<AutoActiveCollectionContainerType>(handler);
```
Comment (jstrzebonski) [inline src/vt/registry/auto/auto_registry.cc:90]: This, and next cases, shouldn't extract handlers ids.

File: src/vt/registry/auto/collection/auto_registry_collection.impl.h [hunk 3]
```diff
@@ -54,7 +54,8 @@ inline AutoActiveCollectionType getAutoHandlerCollection(
   HandlerType const& handler
 ) {
   using ContainerType = AutoActiveCollectionContainerType;
-  return getAutoRegistryGen<ContainerType>().at(handler).getFun();
+  auto const han_id = HandlerManagerType::getHandlerIdentifier(handler);
```
Comment (jstrzebonski) [inline src/vt/registry/auto/collection/auto_registry_collection.impl.h:57]: Verify this file once again - get() extracts id from the handler, but make doesn't use makeHandler(is_auto, is_functor, id), so it might not work as expected.

File: src/vt/registry/auto/rdma/auto_registry_rdma.impl.h [hunk 4]
```diff
@@ -54,7 +54,8 @@ namespace vt { namespace auto_registry {
 
 inline AutoActiveRDMAPutType getAutoHandlerRDMAPut(HandlerType const& handler) {
   using ContainerType = AutoActiveRDMAPutContainerType;
-  return getAutoRegistryGen<ContainerType>().at(handler).getFun();
+  auto const han_id = HandlerManager::getHandlerIdentifier(handler);
+  return getAutoRegistryGen<ContainerType>().at(han_id).getFun();
 }
 
 template <typename MsgT, ActiveTypedRDMAPutFnType<MsgT>* f>
```
Comment (jstrzebonski) [inline src/vt/registry/auto/rdma/auto_registry_rdma.impl.h:61]: Shouldn't this makeHandler call HandlerManager::makeHandler()?

File: src/vt/registry/auto/vc/auto_registry_vc_impl.h [hunk 5]
```diff
@@ -66,7 +66,8 @@ inline HandlerType makeAutoHandlerVC() {
```
Comment (jstrzebonski) [inline src/vt/registry/auto/vc/auto_registry_vc_impl.h:66]: Shouldn't this makeHandler call HandlerManager::makeHandler()?

**Review by jstrzebonski — COMMENTED** [review]
Looks promising, but not quite finished yet. Rework and recheck some of the files / changes considering especially handlers making and ids extraction.

**Comment by jstrzebonski** [comment]
Also, there are two, very very very minor things, that kind of irritates me so much:
1. passing fundamental types by const refs (in this PR's case it's `HandlerType`, which simply is `int64_t`)
2. inconsistency with passing params sometimes by const value and sometimes by value (even if const val could be used).

I know that these are the least of our problems and things on the TODO list, but I wonder if we could, and want to, somehow unify this, going step by step, from time to time. I could start with `HandlerType` in this PR.

What do you thing guys?

**Comment by jstrzebonski** [comment]
OK, so in the end, I've removed redundant const refs.

**Comment by jstrzebonski** [comment]
@lifflander @JacobDomagala @cz4rs @PhilMiller 
Please review the changes.

File: src/vt/vrt/collection/balance/lb_invoke/lb_manager.h [hunk 6]
```diff
@@ -178,12 +178,9 @@ struct LBManager : runtime::component::Component<LBManager> {
   void runLB(LBProxyType base_proxy, PhaseType phase);
 
 private:
-  std::size_t num_invocations_             = 0;
-  std::size_t num_release_                 = 0;
   PhaseType cached_phase_                  = no_lb_phase;
   LBType cached_lb_                        = LBType::NoLB;
   std::function<void()> destroy_lb_        = nullptr;
-  bool synced_in_lb_                       = true;
```
Comment (PhilMiller) [inline src/vt/vrt/collection/balance/lb_invoke/lb_manager.h:186]: Are the changes seen in this file an artifact of some merging or rebasing conflict? They don't seem relevant to the PR.

**Review by PhilMiller — APPROVED** [review]
Looks like some tiny stray changes, but otherwise good.

File: src/vt/vrt/collection/balance/lb_invoke/lb_manager.h (hunk 6 above)
Comment (jstrzebonski) [reply src/vt/vrt/collection/balance/lb_invoke/lb_manager.h:186]: Clang was constantly warning me about unused member vars.
So I got rid of them.

Comment (jstrzebonski) [reply src/vt/vrt/collection/balance/lb_invoke/lb_manager.h:186]: It looked better a week or two ago, but I've recently squashed all the commits

Comment (PhilMiller) [reply src/vt/vrt/collection/balance/lb_invoke/lb_manager.h:186]: No worries, it's fine given the explanation. I hadn't even gotten to looking at the commit log.
